src/exif.c

```diff
@@ -0,0 +1,119 @@
+// Copyright 2022 Google LLC. All rights reserved.
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "avif/internal.h"
+
+#include <stdint.h>
+#include <string.h>
+
+avifResult avifExtractExifTiffHeaderOffset(const avifRWData * exif, uint32_t * offset)
+{
+    const uint8_t tiffHeaderBE[4] = { 'M', 'M', 0, 42 };
+    const uint8_t tiffHeaderLE[4] = { 'I', 'I', 42, 0 };
+    for (*offset = 0; *offset < (exif->size - 4); ++*offset) {
+        if (!memcmp(&exif->data[*offset], tiffHeaderBE, 4) || !memcmp(&exif->data[*offset], tiffHeaderLE, 4)) {
+            return AVIF_RESULT_OK;
+        }
+    }
+    // Couldn't find the TIFF header
+    return AVIF_RESULT_INVALID_EXIF_PAYLOAD;
+}
+
+avifResult avifExtractExifOrientationToIrotImir(avifImage * image)
+{
+    uint32_t offset;
+    const avifResult result = avifExtractExifTiffHeaderOffset(&image->exif, &offset);
+    if (result != AVIF_RESULT_OK) {
+        // Couldn't find the TIFF header
+        return result;
+    }
+
+    avifROData raw = { image->exif.data + offset, image->exif.size - offset };
+    const avifBool littleEndian = (raw.data[0] == 'I');
+    avifROStream stream;
+    avifROStreamStart(&stream, &raw, NULL, NULL);
+
+    // TIFF Header
+    uint32_t offsetToNextIfd;
+    if (!avifROStreamSkip(&stream, 2 + 2) ||                                       // Byte Order ("II" or "MM" then "42")
+        !avifROStreamReadU32Endianness(&stream, &offsetToNextIfd, littleEndian) || // Offset to 0th IFD
+        offsetToNextIfd < 2 + 2 + 4) {
```

Comment:
> Is the `offsetToNextIfd < 2 + 2 + 4` check necessary?

It was an easy bitstream-compliance check. I removed it.

> You don't do this check after the `avifROStreamReadU32(&stream, &offsetToNextIfd)` call at line 114.

No because I do not know if the chunks are always in order so the check to perform is not obvious.

> NOTE: It is important to check that `offsetToNextIfd` does not go past the end of Exif metadata. This is done implicitly by the `avifROStreamSetOffset(&stream, offsetToNextIfd)` call at line 45.

Acknowledged.